README.rst
.. These are examples of badges you might want to add to your README:
   please update the URLs accordingly

    .. image:: https://api.cirrus-ci.com/github/<USER>/mastrsql.svg?branch=main
        :alt: Built Status
        :target: https://cirrus-ci.com/github/<USER>/mastrsql
    .. image:: https://readthedocs.org/projects/mastrsql/badge/?version=latest
        :alt: ReadTheDocs
        :target: https://mastrsql.readthedocs.io/en/stable/
    .. image:: https://img.shields.io/coveralls/github/<USER>/mastrsql/main.svg
        :alt: Coveralls
        :target: https://coveralls.io/r/<USER>/mastrsql
    .. image:: https://img.shields.io/pypi/v/mastrsql.svg
        :alt: PyPI-Server
        :target: https://pypi.org/project/mastrsql/
    .. image:: https://img.shields.io/conda/vn/conda-forge/mastrsql.svg
        :alt: Conda-Forge
        :target: https://anaconda.org/conda-forge/mastrsql
    .. image:: https://pepy.tech/badge/mastrsql/month
        :alt: Monthly Downloads
        :target: https://pepy.tech/project/mastrsql


.. image:: https://img.shields.io/badge/-PyScaffold-005CA0?logo=pyscaffold
    :alt: Project generated with PyScaffold
    :target: https://pyscaffold.org/

|

===============
mastrsql
===============


    A python package to download the german Marktstammdatenregister (short: MaStR) and save it as a local PostgreSQL database.


The german *Federal Network Agency for Electricity, Gas, Telecommunications, 
Post and Railway* (Bundesnetzagentur) publishes their data set of all electricity and gas producers openly 
on their website_. This data set is called *Markstammdatenregister* (short: MaStR). As a relatively new feature (as of 2021), 
they offer a complete `data download`_ in zipped xml format 
besides their API_. The scope of the mastrsql package is to offer an easy and automated download of the MaStR into a local 
PostgreSQL database. It is planned to expand the package to also include an update function, that only downloads new entries
which are not yet saved in the local database via the API. If you're looking for other features, also check out `open-Mastr`_

Quick Start
============
Install the package from PyPI (as soon as it is published there).

.. code:: bash

    pip install mastrsql


Import the Mastr class and define a Mastr object.

>>> from mastrsql.mastr import Mastr 

>>> database = Mastr()

Download the latest version of the zipped MaStR in xml format.

>>> database.download()

Read the downloaded files into a PostgreSQL database.

>>> database.to_sql()

For extracting information from the local PostgreSQL database we refer to other packages, for example you can use `pandas.read_sql`_.

Note
====

This project has been set up using PyScaffold 4.1.1. For details and usage
information on PyScaffold see https://pyscaffold.org/.


.. _website: https://www.marktstammdatenregister.de/MaStR
.. _data download: https://www.marktstammdatenregister.de/MaStR/Datendownload 
.. _API: https://www.marktstammdatenregister.de/MaStRHilfe/subpages/webdienst.html
.. _pandas.read_sql: https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html
.. _open-Mastr: https://open-mastr.readthedocs.io/en/dev/



CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- First published version
- Added a Mastr class
- Added the Mastr class function initialize
- Added the Mastr class function to_sql
- Added util functions


tests/conftest.py
"""
    Dummy conftest.py for mastrsql.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


tests/mastr_test.py
from mastrsql.mastr import Mastr
import pytest
import os
import pandas as pd


def test_Mastr_init():
    with pytest.raises(AssertionError):
        user_credentials = "This is a string"
        mastr = Mastr(user_credentials=user_credentials)
 
    mastr = Mastr()
    assert type(mastr.today) == str
    assert type(mastr.url) == str
    assert "marktstammdatenregister" in mastr.url



def test_Mastr_download():
    mastr = Mastr()
    mastr.download()
    assert os.path.exists(mastr.save_zip_path) 
    assert os.path.getsize(mastr.save_zip_path) > 900000000

def test_Mastr_to_sql():
    exclude_tables=["anlageneegsolar", "einheitensolar", "lokationen", "marktakteure", "netzanschlusspunkte","marktrollen"]
    include_tables=["netze","einheitengaserzeuger","marktrollen"]
    mastr = Mastr()
    mastr.to_sql(exclude_tables=exclude_tables)
    

    df=pd.read_sql("netze", con = mastr.engine)
    assert type(df)==pd.core.frame.DataFrame
    assert len(df) >= 100

    df_2=pd.read_sql("einheitengaserzeuger", con = mastr.engine)
    assert type(df_2)==pd.core.frame.DataFrame
    assert len(df_2) >= 100

    mastr.to_sql(include_tables=include_tables)
    df_3=df_2=pd.read_sql("marktrollen", con = mastr.engine)
    assert type(df_3)==pd.core.frame.DataFrame
    assert len(df_3) >= 100



tests/utils_test.py
from mastrsql.utils import get_url, correction_of_metadata
import pandas as pd
import numpy as np


def test_get_url():
    assert type(get_url()) == str


def test_correction_of_metadata():
    df = pd.DataFrame(
        {
            "Postleitzahl": [10000, "10001"],
            "Laengengrad": ["799.3",33.33],
            "Inbetriebnahmedatum": ["20201003","18001010"]
        }
    )
    sql_tablename = "einheitengasspeicher"
    df_2, sql_dtype_dict = correction_of_metadata(df,sql_tablename)

    assert type(df_2) == pd.core.frame.DataFrame
    assert len(df_2) == 2
    assert type(df_2.Postleitzahl[1]) == np.int32
    assert type(df_2.Inbetriebnahmedatum[1]) == pd._libs.tslibs.timestamps.Timestamp
    assert type(sql_dtype_dict) == dict
    assert len(sql_dtype_dict) == 3



src/mastrsql/utils.py
import requests
import os
from clint.textui import progress
import time
import sqlalchemy
import pandas as pd
import psycopg2
from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
from bs4 import BeautifulSoup
import pdb
import dateutil
from mastrsql.metadata import metadata_dict


def get_url():
    """Get the url of the latest MaStR file from markstammdatenregister.de.

    The file and the corresponding url are updated once per day.
    The url has a randomly generated string appended, so it has to be
    grabbed from the marktstammdatenregister.de homepage.
    For further details visit https://www.marktstammdatenregister.de/MaStR/Datendownload
    """
    html = requests.get("https://www.marktstammdatenregister.de/MaStR/Datendownload")
    soup = BeautifulSoup(html.text, "lxml")
    # find the download button element on the website
    element = soup.find_all("a", "btn btn-primary text-right")[0]
    # extract the url from the html element
    url = str(element).split('href="')[1].split('" title')[0]
    return url


def download_from_url(url, save_path):
    """Downloads the zipped MaStR.

    Parameters
    -----------
    url : str
        The url where the MaStR can be downloaded.
    save_path: str
        The path where the downloaded MaStR zipped folder will be saved.
    """
    print("Download has started, this can take several minutes.")
    time_a = time.perf_counter()
    r = requests.get(url, stream=True)
    with open(save_path, "wb") as zfile:
        total_length = int(7400 * 1024 * 1024)
        for chunk in progress.bar(
            r.iter_content(chunk_size=1024 * 1024),
            # chunk size of 1024 * 1024 needs 9min 11 sek = 551sek
            # chunk size of 1024 needs 9min 11 sek as well
            expected_size=(total_length / 1024 / 1024),
        ):
            if chunk:
                zfile.write(chunk)
                zfile.flush()
    time_b = time.perf_counter()
    print("Download is finished. It took %s seconds." % (time_b - time_a))


def correction_of_metadata(df, sql_tablename):
    """Changes data types of Dataframe columns according to predefined metadata.

    Parameters
    -----------
    df : pandas.DataFrame
        DataFrame of MaStR tables read from xml files
    sql_tablename : str
        Name of the table in the PostrgreSQL database

    Returns
    --------
    df : pandas.DataFrame
        DataFrame of MaStR with corrected metadata
    sql_dtype_dict : dict
        Dictionary of column name / data type pairs that are needed
        for the sql dump of the DataFrame
    """

    transform_dtypes_sql = {
        "str": sqlalchemy.types.Text,
        "int": sqlalchemy.types.Integer,
        "float": sqlalchemy.types.Float,
        "datetime64": sqlalchemy.types.DateTime,
        "datetime64[D]": sqlalchemy.types.Date,
    }

    dtype_dict = metadata_dict[sql_tablename]
    # The loaded dtype_dict usually specifies more columns than there are
    # columns in the df. We have to create a dtype_dict_for_df that only
    # has keys that are columns of the list.
    dtype_dict_for_df = {}
    sql_dtype_dict = {}
    for column in df.columns:
        dtype_dict_for_df[column] = dtype_dict[column]
        sql_dtype_dict[column] = transform_dtypes_sql[dtype_dict[column]]
    try:
        df = df.astype(dtype_dict_for_df)
    except pd._libs.tslibs.np_datetime.OutOfBoundsDatetime:
        # Some datetimes ly outside the interval of 677-09-21 and
        # 2262-04-11, which causes an error.
        # With pd.to_datetime, those values are set to NaT
        for key in dtype_dict_for_df:
            if (
                dtype_dict_for_df[key] == "datetime64[D]"
                or dtype_dict_for_df[key] == "datetime64"
            ):
                df[key] = pd.to_datetime(df[key], errors="coerce")
    except (pd.errors.IntCastingNaNError, dateutil.parser._parser.ParserError):
        # Errors not related to datetimes are ignored.
        # Then the column where the error occured will not be changed,
        # but all other columns change their data type
        df = df.astype(dtype_dict_for_df, errors="ignore")
    return df, sql_dtype_dict



def initialize_database(user_credentials, postgres_standard_credentials = {}):
    """Create new PostgreSQL database if it doesn't exist yet.

    Parameters
    ------------
    user_credentials : dict
        Dictionary of credentials for the database.

    postgres_standard_credentials : dict, default {}
        Dictionary of credentials for the database initially created
        when installing PostgresSQL. Possible keys are
        "dbname" default "postgres", "user" default "postgres",
        "password" default "postgres", "host" default "localhost",
        "port" default "5432". The given default values are insterted
        into the postgres_standard_credentials if they are not given
        by the user.

    """
    
    postgres_standard_credentials_temp = {
        "dbname": "postgres",
        "user": "postgres",
        "password": "postgres",
        "host": "localhost",
        "port": "5432",
        }

    postgres_standard_credentials = {**postgres_standard_credentials_temp, **postgres_standard_credentials}    

    try:
        con = psycopg2.connect(
            dbname=postgres_standard_credentials["dbname"],
            user=postgres_standard_credentials["user"],
            password=postgres_standard_credentials["password"],
            host=postgres_standard_credentials["host"],
            port=postgres_standard_credentials["port"],
        )
        con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
        cursor = con.cursor()
        name_database = user_credentials["dbname"]
        cursor.execute(f"CREATE DATABASE {name_database};")
        cursor.close()
        con.close()
    except psycopg2.errors.DuplicateDatabase:
        print(f"Using existing {name_database} database in PostgreSQL.")


def handle_xml_syntax_error(data, err):
    """Deletes entries that cause an xml syntax error and produces DataFrame.

    Parameters
    -----------
    data : bytes
        Unzipped xml data
    err : ErrorMessage
        Error message that appeared when trying to use pd.read_xml on invalid xml file.

    Returns
    ----------
    df : pandas.DataFrame
        DataFrame which is read from the changed xml data.
    """
    
    # Actually it is unclear if there are still invalid xml files in the recent MaStR.

    wrong_char_position = int(str(err).split()[-4])
    decoded_data = data.decode("utf-16")
    loop_condition = True

    shift = 0
    while loop_condition:
        evaluated_string = decoded_data[wrong_char_position + shift]
        if evaluated_string == ">":
            start_char = wrong_char_position + shift + 1
            break
        else:
            shift -= 1
    loop_condition_2 = True
    while loop_condition_2:
        evaluated_string = decoded_data[start_char]
        if evaluated_string == "<":
            break
        else:
            decoded_data = decoded_data[:start_char] + decoded_data[start_char + 1 :]
    df = pd.read_xml(decoded_data)
    print("One invalid xml expression was deleted.")
    return df


src/python_mastrsql/skeleton.py
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
``[options.entry_points]`` section in ``setup.cfg``::

    console_scripts =
         fibonacci = python_mastrsql.skeleton:run

Then run ``pip install .`` (or ``pip install -e .`` for editable mode)
which will install the command ``fibonacci`` inside your current environment.

Besides console scripts, the header (i.e. until ``_logger``...) of this file can
also be used as template for Python modules.

Note:
    This skeleton file can be safely removed if not needed!

References:
    - https://setuptools.readthedocs.io/en/latest/userguide/entry_point.html
    - https://pip.pypa.io/en/stable/reference/pip_install
"""

import argparse
import logging
import sys

from python_mastrsql import __version__

__author__ = "Florian Kotthoff"
__copyright__ = "Florian Kotthoff"
__license__ = "MIT"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from python_mastrsql.skeleton import fib`,
# when using this Python module as a library.


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for _i in range(n - 1):
        a, b = b, a + b
    return a


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="python-mastrsql {ver}".format(ver=__version__),
    )
    parser.add_argument(dest="n", help="n-th Fibonacci number", type=int, metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )


def main(args):
    """Wrapper allowing :func:`fib` to be called with string arguments in a CLI fashion

    Instead of returning the value from :func:`fib`, it prints the result to the
    ``stdout`` in a nicely formatted message.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--verbose", "42"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m python_mastrsql.skeleton 42
    #
    run()


src/mastrsql/mastr.py
import numpy as np
import os
from os.path import expanduser
from zipfile import ZipFile
import sqlalchemy
from sqlalchemy import create_engine
import pandas as pd
import pdb
import psycopg2
import lxml
import shutil
from mastrsql.utils import (
    get_url,
    download_from_url,
    correction_of_metadata,
    handle_xml_syntax_error,
    initialize_database,
)
from datetime import date


class Mastr:
    """Mirrors the MaStR (Marktstammdatenregister) to a PostrgreSQL data base.

    Parameters
    ------------
    user_credentials : dict, default {}
        Dictionary of credentials for the MaStR-database. Possible keys are
        "dbname", "user", "password", "host", "port". Key-Value pairs that are not
        explicitly given will be set to 
        "dbname" : "mastrsql",
        "user": "postgres",
        "password": "postgres",
        "host": "localhost",
        "port": "5432"

    postgres_standard_credentials : dict, default {}
        Dictionary of credentials for the first log in to postgres. Possible keys are
        "dbname", "user", "password", "host", "port". Key-Value pairs that are not
        explicitly given will be set to 
        "dbname" : "postgres",
        "user": "postgres",
        "password": "postgres",
        "host": "localhost",
        "port": "5432"

    Returns
    --------
    out : mastrsql.mastr.Mastr
        An Mastr object that represents the MaStR database.
    """


    def __init__(self, user_credentials={}, postgres_standard_credentials={}):
        assert type(user_credentials) == dict, "Variable user_credentials has to be a python dictionary!"
        
        #self.today = date.today().strftime("%Y%m%d")
        self.today = "20211026"
        self.url = get_url()
        self.save_path = os.path.join(
            expanduser("~"),
            ".mastrsql",
            "data",
        )

        standard_credentials = {
        "dbname": "mastrsql",
        "user": "postgres",
        "password": "postgres",
        "host": "localhost",
        "port": "5432",
        }
        self.user_credentials = {**standard_credentials, **user_credentials}
    
    
        self.save_zip_path = os.path.join(
            self.save_path, "Gesamtdatenexport_%s.zip" % self.today
        )

        initialize_database(self.user_credentials,postgres_standard_credentials=postgres_standard_credentials)
        create_engine_string=f"postgresql+psycopg2://postgres:{self.user_credentials['user']}@{self.user_credentials['host']}:{self.user_credentials['port']}/{self.user_credentials['dbname']}"
        self.engine = create_engine(create_engine_string)

    def download(self, delete_old=True):
        """Downloads the latest MaStR zipped file to HOME/.mastrsql/data
        
        Parameters
        ------------
        delete_old : boolean, default True
            If delete_old is True, older versions of the downloaded zipped
            MaStR will be deleted if found in self.save_zip_path
        """
        if os.path.exists(self.save_zip_path):
            print("MaStR already downloaded.")
        else:
            if delete_old:
                shutil.rmtree(self.save_path)
            print("MaStR is downloaded to %s" % self.save_path)
            os.makedirs(self.save_path, exist_ok=True)
            # download data from url
            download_from_url(self.url, self.save_zip_path)

    def to_sql(self,include_tables=None,exclude_tables=None):
        """Writes the local zipped MaStR to a PostgreSQL database.
        
        Parameters
        ------------
        include_tables : list, default None
            List of tables from the Marktstammdatenregister that should be written into
            the database. Elements of include_tables are lower case strings without "_" and index. 
            It is possible to include any table from the zipped local MaStR folder (see MaStR.initialize()). 
            Example: If you do want to write the data from files "AnlagenEegSolar_*.xml" to a table 
            in your database (where * is any number >=1), write the element "anlageneegsolar" into the 
            include_tables list. Elements of the list that cannot be matched to tables of the MaStR are ignored.
            If include_tables is given, only the tables listed here are written to the database.

        exclude_tables : list, default None
            List of tables from the Marktstammdatenregister that should NOT be written into
            the database. Elements of exclude_tables are lower case strings without "_" and index. 
            It is possible to exclude any table from the zipped local MaStR folder (see MaStR.initialize()). 
            Example: If you do not want to write the data from files "AnlagenEegSolar_*.xml" to a table 
            in your database (where * is any number >=1), write the element "anlageneegsolar" into the 
            exclude_tables list. Elements of the list that cannot be matched to tables of the MaStR are ignored.

        """

        # exclude_count_reference is important for checking which files are written to the SQL database
        # if it is 1, all files from the exclude_tables_reference are included to be written to the databse,
        # if it is 0, all files from the exclude_tables_reference are excluded.

        if include_tables:
            exclude_count_reference = 1
            exclude_tables_reference=include_tables
        elif exclude_tables:
            exclude_count_reference = 0
            exclude_tables_reference=exclude_tables
        else:
            exclude_count_reference = 0
            exclude_tables_reference=[]


        with ZipFile(self.save_zip_path, "r") as f:
            for name in f.namelist():
                # sql tablename is the beginning of the filename without the number in lowercase
                sql_tablename = name.split("_")[0].split(".")[0].lower()
                
                # check whether the table exists with current data and append new data or whether to overwrite the existing table
                

                exclude_count = exclude_tables_reference.count(sql_tablename)
                if exclude_count == exclude_count_reference:
                    index_for_printed_message=1
                    if (
                    name.split(".")[0].split("_")[-1] == "1"
                    or len(name.split(".")[0].split("_")) == 1
                    ):
                        if_exists = "replace"
                        print("New table %s is created in the PostgreSQL database." %sql_tablename)
                    else:
                        if_exists = "append"
                        print(f"File {index_for_printed_message} is parsed.")
                        index_for_printed_message+=1
                    
                    data = f.read(name)
                    save_path_metadata = os.path.join(expanduser("~"),".mastrsql","metadata")
                    try:
                        df = pd.read_xml(data, encoding="UTF-16", compression="zip")
                        df, sql_dtype_dict = correction_of_metadata(
                            df, sql_tablename
                        )

                    except lxml.etree.XMLSyntaxError as err:
                        df = handle_xml_syntax_error(data, err)
                        df, sql_dtype_dict = correction_of_metadata(
                            df, sql_tablename, save_path_metadata
                        )
                    continueloop = True

                    """Some files introduce new columns for existing tables. 
                    If this happens, the error from writing entries into non-existing columns is caught and the column is created."""
                    while continueloop:
                        try:
                            df.to_sql(
                                sql_tablename,
                                self.engine,
                                if_exists=if_exists,
                                dtype=sql_dtype_dict,
                            )
                            continueloop = False
                        except sqlalchemy.exc.ProgrammingError as err:
                            missing_column = str(err).split("«")[0].split("»")[1]
                            con = psycopg2.connect(
                                "dbname=mastrsql user=postgres password='postgres'"
                            )
                            cursor = con.cursor()
                            execute_message = 'ALTER TABLE %s ADD "%s" text NULL;' % (
                                sql_tablename,
                                missing_column,
                            )
                            cursor.execute(execute_message)
                            con.commit()
                            cursor.close()
                            con.close()
                        except sqlalchemy.exc.DataError as err:
                            delete_entry = str(err).split("«")[0].split("»")[1]
                            print(delete_entry)
                            df = df.replace(delete_entry, np.nan)



if __name__ == "__main__":
    database = Mastr()
    #database.download()
    exclude_tables=["anlageneegsolar", "einheitensolar", "lokationen", "marktakteure", "netzanschlusspunkte"]
    database.to_sql(exclude_tables=exclude_tables)
